### ISSUE TITLE
Deciding the model to train (and the base code to profile!)

### ISSUE THREAD

**Issue by SeanNaren** [SeanNaren/min-LLM#5]
There is an epic paper from DeepMind: [Training Compute-Optimal Large Language Models](https://arxiv.org/pdf/2203.15556.pdf)

TDLR: Large Transformer models are probably not being trained optimally (datasets are smaller than optimal and training times too short).

<img width="500" alt="Screenshot 2022-04-08 at 22 26 29" src="https://user-images.githubusercontent.com/6707363/162627527-f1358c48-3cf8-41e7-a933-c608645a5ade.png">

Deciding which model to train comes down to how many FLOP per second I can get to on one multi-GPU machine, and how patient I am it seems! As a side-effect, the more efficient the operations, the quicker training will be.

After speaking to @blefaudeux, reading [this](https://horace.io/brrr_intro.html) amazing blog post by @Chillee and looking at what was done in the BigScience configuration selection process with @stas00s' [benchmarks](https://github.com/bigscience-workshop/bigscience/blob/master/train/tr11-176B-ml/chronicles-prequel.md#best-tflops) I think I can use DeepSpeed's Profiler's `fwd FLOPS per GPU = fwd flops per GPU / fwd latency` metric to try out various precisions and model configurations.

Work has begun in #4!

**Comment by blefaudeux** [comment]
Adding two items to check out if you want:
- we just landed a mem optimized attention (not training yet, but will come) in xformers, using the Rabe algo, the impact on memory use is just massive, better deal than checkpointing for now
- nvidia's tinyNN module is really really fast, hitting above its weight (fused MLP). It will be limited in terms of dimensions but if you can use it, could be a game changer with respect to training time

**Comment by SeanNaren** [comment]
Thanks as always @blefaudeux! 

- The Rabe algo looks really cool, seemed to get a lot of flack on reddit but seems the savings are real! Will wait for the BW to be implemented and will definitely benchmark it :)
- tinyNN looks really cool, but seems maximum neurons is [128](https://github.com/NVlabs/tiny-cuda-nn/blob/master/DOCUMENTATION.md#fully-fused-mlp). I may be wrong so will double check, but I think this might not be feasible unfortunately :(

**Comment by blefaudeux** [comment]
>     * tinyNN looks really cool, but seems maximum neurons is [128](https://github.com/NVlabs/tiny-cuda-nn/blob/master/DOCUMENTATION.md#fully-fused-mlp). I may be wrong so will double-check, but I think this might not be feasible unfortunately :(

oh wow, it's pretty small indeed :) you can always use a couple of them but that defeats the purpose. They do offer another variant though, which can take any size, [see](https://github.com/NVlabs/tiny-cuda-nn/blob/master/DOCUMENTATION.md#cutlass-mlp)

**Comment by SeanNaren** [comment]
Didn't catch that, nice!! Will play around with it for sure

**Comment by Chillee** [comment]
@blefaudeux I mean... the sqrt(N) memory attention *is* just checkpointing - it's just a particular variant of checkpointing.

**Comment by blefaudeux** [comment]
> @blefaudeux I mean... the sqrt(N) memory attention _is_ just checkpointing - it's just a particular variant of checkpointing.

you save memory for inference, which is not really checkpointing related ? the backward part is checkpointing yes, but within the attention mechanism, which is not trivially doable with the checkpoint wrappers out there

**Comment by blefaudeux** [comment]
> 

I mean, wake me up when you can save this type of memory during inference by checkpointing (or any for that matter :D) https://github.com/facebookresearch/xformers/pull/267

**Comment by blefaudeux** [comment]
@Chillee maybe that I'm missing something actually, can you walk me through how you would do this with checkpointing ? checkpoint before the attention ? you still get the attention memory bubble anyway, so if this is the limiting factor it does not help that much, but you do save on the activations (that's a /2 win, no sqrt(N)). Am I missing something ?

**Comment by blefaudeux** [comment]
> Thanks as always @blefaudeux!
> 
>     * The Rabe algo looks really cool, seemed to get a lot of flack on reddit but seems the savings are real! Will wait for the BW to be implemented and will definitely benchmark it :)
> 
>     * tinyNN looks really cool, but seems maximum neurons is [128](https://github.com/NVlabs/tiny-cuda-nn/blob/master/DOCUMENTATION.md#fully-fused-mlp). I may be wrong so will double-check, but I think this might not be feasible unfortunately :(
> 
> 
> If you have any more tricks/ideas throw them my way! Additional point from offline discussions for book-keeping purposes.
> 
>     * Might be worth having a look at BlockSparse (trying a causal mask at the least).
> 
> 
> Optimization of training time is important! I think it's worth doing the necessary tests to benchmark various configurations/model changes.

oh wow you meant this https://www.reddit.com/r/MachineLearning/comments/rjgbxc/p_memory_efficient_attention_selfattention_does/ ? I think that some of the comments are just factually wrong. 

Yes, there are other mechanisms which scale similarly, but they approximate, apples and oranges. They are definitely interesting (it's like half the content on xformers..), but apples and oranges, it does not devaluate the paper one bit. 

No to my understanding you cannot get anything remotely similar with checkpointing (I'd love to be proven wrong, looks like the commenters on this reddit thread did not follow up)

**Comment by SeanNaren** [comment]
> oh wow you meant this https://www.reddit.com/r/MachineLearning/comments/rjgbxc/p_memory_efficient_attention_selfattention_does/ ? I think that some of the comments are just factually wrong.
> 
> Yes, there are other mechanisms which scale similarly, but they approximate, apples and oranges. They are definitely interesting (it's like half the content on xformers..), but IMO it does not devaluate the paper one bit.

Yeah seemed pretty harsh with little follow-up, but would be curious how this potentially folds into some variant of checkpointing (I have a hard time seeing it as well)

**Comment by Chillee** [comment]
> you save memory for inference, which is not really checkpointing related ? the backward part is checkpointing yes, but within the attention mechanism, which is not trivially doable with the checkpoint wrappers out there

Yes, to clarify, it's 1. chunking the attention computation, and 2. standard sqrt(N) checkpointing. Similarly, you could lower the memory usage down to log(N) (while evaluating the forwards pass log(N) times) by using the standard binomial checkpointing scheme.

**Comment by Chillee** [comment]
> how do you checkpoint within softmax(QKtV)

That's the chunking :) You do need a special trick to delay the softmax division.

You're doing something along the lines of 

```
1. Compute first row of attention.
2. Compute second row of attention
3. Compute third row of attention
...
```
and then checkpoint that.

**Comment by blefaudeux** [comment]
> > how do you checkpoint within softmax(QKtV)
> 
> That's why you need the chunking :) You also need a special trick to delay the softmax division.
> 
> You're doing something along the lines of
> 
> ```
> 1. Compute first row of attention.
> 2. Compute second row of attention
> 3. Compute third row of attention
> ...
> ```
> 
> and then checkpoint that.

you're describing the paper here, and AFAIK this technique did not exist before the paper, so I guess we agree ?

**Comment by blefaudeux** [comment]
> @blefaudeux I mean... the sqrt(N) memory attention _is_ just checkpointing - it's just a particular variant of checkpointing.

this made it sound like the paper was just checkpointing, well it's one of the parts used (for the backward, again this is not used for the FW), the rest is new. You cannot seriously do this with [pytorch checkpointing](https://pytorch.org/docs/stable/checkpoint.html?highlight=checkpoint#torch.utils.checkpoint.checkpoint) for instance. "Just a particular variant of checkpointing" means "included in the superset of checkpointing" which to my understanding is not exact

**Comment by Chillee** [comment]
> you're describing the paper here, and AFAIK this technique did not exist before the paper, so I guess we agree ?

My point is that both 1. chunking, and 2. checkpointing are standard approaches to reduce memory. Anytime you have a large intermediate, a natural thing to do is to compute the intermediate in chunks. But yes, I agree that AFAIK, I haven't seen anybody apply it specifically to transformers.

**Comment by blefaudeux** [comment]
yeah well, so the contributions from the paper are the numerical trick to update the softmax normalization a posteriori, and thinking of how to apply that to the attention mechanism (related of course). Does not sound much less impressive than something like the dropout paper, is it ? :D Anyway, I think we agree ! I would definitely include this in the bag of tricks for a "Small Science" experiment, anything that allows you to cram as much as possible within a node it good to have and the speed penalty is fairly small

**Comment by SeanNaren** [comment]
Closing this, as a lot of changes have been merged. After some experiments run offline, I'm happy with where the architecture has landed!
